src/cluster/mcl.cpp

```diff
@@ -0,0 +1,443 @@
+	std::chrono::high_resolution_clock::time_point t = std::chrono::high_resolution_clock::now();
+	if( r - (int) r == 0 ){
+		// TODO: at some r it may be more beneficial to diagnoalize in and only take the exponents of the eigenvalues
+		*out = *in;
+		for(uint32_t i=1; i<r; i++){ 
+			(*out) = (*out)*(*in);
+		}
+	}
+	else{ 
+		throw runtime_error(" Eigen does not provide an eigenvalue solver for sparse matrices");
+	}
+	*out = out->pruned();
+	sparse_exp_time += (double) std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - t).count() / 1000.0;
+}
+
+void MCL::get_exp(Eigen::MatrixXf* in, Eigen::MatrixXf* out, float r){
+	// TODO: at some r it may be more beneficial to diagnoalize in and only take the exponents of the eigenvalues
+	std::chrono::high_resolution_clock::time_point t = std::chrono::high_resolution_clock::now();
+	if(r - (int) r == 0){
+		*out = *in;
+		for(uint32_t i=1; i<r; i++){ 
+			(*out) *= (*in);
+		}
+		dense_int_exp_time += (double) std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - t).count() / 1000.0;
+	}
+	else{ 
+		// TODO: check whether the matrix is self-adjoint and use SelfAdjointEigenSolver instead
+		// TODO: try and get out->noalias() = in->pow(r); to work (unsupported! http://eigen.tuxfamily.org/dox/unsupported/group__MatrixFunctions__Module.html).
+		Eigen::EigenSolver<Eigen::MatrixXf> solver(*in);
+		Eigen::MatrixXcf d = solver.eigenvalues().asDiagonal();
+		for(uint32_t idiag = 0; idiag<d.rows(); idiag++){
+			d(idiag, idiag) = pow(d(idiag, idiag), r);
+		}
+		Eigen::MatrixXcf V = solver.eigenvectors();
+		double thr = 0.5 * abs(V.determinant());
+		if( thr > std::numeric_limits<float>::epsilon() ){
+			out->noalias() = (V * d.real() * V.inverse()).real();
+		}
+		dense_gen_exp_time += (double) std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - t).count() / 1000.0;
+	}
+}
+
+void MCL::get_gamma(Eigen::SparseMatrix<float>* in, Eigen::SparseMatrix<float>* out, float r){
+	std::chrono::high_resolution_clock::time_point t = std::chrono::high_resolution_clock::now();
+	vector<Eigen::Triplet<float>> data;
+	for (uint32_t k=0; k<in->outerSize(); ++k){
+		float colSum = 0.0f;
+		for (Eigen::SparseMatrix<float>::InnerIterator it(*in, k); it; ++it){
+			colSum += pow(it.value(),r);
+		}
+		for (Eigen::SparseMatrix<float>::InnerIterator it(*in, k); it; ++it){
+			data.emplace_back(it.row(), it.col(), pow(it.value(),r) / colSum);
+		}
+	}
+	out->setFromTriplets(data.begin(), data.end(), [] (const float&, const float &b) { return b; });
+	*out = out->pruned(1.0, std::numeric_limits<float>::epsilon());
+	sparse_gamma_time += (double) std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - t).count() / 1000.0;
+}
+
+void MCL::get_gamma(Eigen::MatrixXf* in, Eigen::MatrixXf* out, float r){
```

Comment:
One may consider using Eigen's builtin functions here instead, such as the pow() method of the matrix, as well as the colwise() method, see here: https://eigen.tuxfamily.org/dox/group__TutorialReductionsVisitorsBroadcasting.html

This could be faster due to the SIMD acceleration but I'm not sure if it would make a big difference here, but could be worth a try.

Reply:
I have considered using pow, but it is not supported, therefore I have implemented the exp method. The reason for why it is not supported is that the matrix exponential is ill-defined if the eigenvector matrix is not invertible. To circumvent that I have now restricted the input to whole numbers. An alternative approach would be to _always_ use a symmetric similarity matrix, the the eigenvector matrix is unitary and only simple transposition is required to get the inverse. I will go with restricting the input to integral values for now, but if you have any preferences, please let me know.

Reply:
As for the colwise, as far as I can see, I would need two calls to colwise (reduction and broadcasting), while iterating through will only give me a single pass through the data. I can implement a version for performance testing.